### ISSUE TITLE
Reuse memory buffers across iterations 

### ISSUE THREAD

**Issue by zhisbug** [alpa-projects/alpa#147]
To support gradient accumulation, we need to allocate extra buffers where we can accumulate gradients, in addition to buffers that hold gradients generated on-the-fly.

We currently re-allocate these buffers at the start of each training iteration. We are thinking of only allocating them once at the start of the 1st iteration, and reusing these buffers across iterations because re-allocation seems to cause some overheads, see below evidence:

![image](https://user-images.githubusercontent.com/1654062/139382114-59279b69-ae99-4915-b2fa-4edd062e283a.png)

Zoom-in version:
![image](https://user-images.githubusercontent.com/1654062/139382188-75e4884b-3c64-4cf4-85cb-ab1693b42949.png)

The way to achieving this is to replace ALLOC instruction using memset.

Some related concerns is to properly synchronize the memset and allocate so to avoid race conditions.
